internal/app.go

```diff
@@ -205,86 +209,102 @@ func updateSecret(secretFile string, newContent []byte) (bool, error) {
 	return true, nil
 }
 
-func (a *App) extract(crypto CryptoHandler) error {
+func (a *App) extract(crypto CryptoHandler, config configSnapshot) error {
 	if _, err := os.Stat(a.SecretsDir); err != nil {
 		return err
 	}
-	config, err := Unmarshall(crypto, a.EncryptedConfig)
-	if err != nil {
-		return err
-	}
 
-	for fname, cfgFile := range config {
+	all_fname := make(map[string]bool)
+	for fname, cfgFile := range config.next {
 		log.Printf("Extracting %s", fname)
+		all_fname[fname] = true
 		fullpath := filepath.Join(a.SecretsDir, fname)
 		changed, err := updateSecret(fullpath, []byte(cfgFile.Value))
 		if err != nil {
 			return err
 		}
-		if changed && len(cfgFile.OnChanged) > 0 {
-			log.Printf("Running on-change command for %s: %v", fname, cfgFile.OnChanged)
-			cmd := exec.Command(cfgFile.OnChanged[0], cfgFile.OnChanged[1:]...)
-			cmd.Env = append(os.Environ(), "CONFIG_FILE="+fullpath)
-			cmd.Stdout = os.Stdout
-			cmd.Stderr = os.Stderr
-			if err := cmd.Run(); err != nil {
-				log.Printf("Unable to run command: %v", err)
-			}
+		if changed {
+			runOnChanged(fname, fullpath, cfgFile.OnChanged)
+		}
+	}
+
+	// Now, watch for file removals (compare with a previous version if present)
+	if config.prev == nil {
+		return nil
+	}
+	for fname, cfgFile := range config.prev {
+		if _, ok := all_fname[fname]; ok {
+			continue
 		}
+		log.Printf("Removing %s", fname)
+		fullpath := filepath.Join(a.SecretsDir, fname)
+		if err := os.Remove(fullpath); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		runOnChanged(fname, fullpath, cfgFile.OnChanged)
 	}
+
 	return nil
 }
 
 func (a *App) Extract() error {
 	_, crypto := createClient(a.sota)
 	defer crypto.Close()
-	return a.extract(crypto)
-}
 
-func close(c io.Closer, name string) {
```

Comment:
The `defer close(...)` lines you've added all return possible error values. I thought the linter `make check` runs caught them. Anyway commit 3eba21e79fae681ea4e633a2cbf067a859cbc420 added this to make a linter happy. I think we should probably keep using it so we can log unexpected errors.

Reply:
I was actually quite surprised when I first seen that we defer Close for a writable file, so I thought Go is different than Python i.e. it doesn't need to flush uncommitted writes during a Close.
But per this article it appears that it's the same:
https://www.joeshaw.org/dont-defer-close-on-writable-files/

So, I'm going to Close a file properly in the next patch set and treat any error just like a failed Write call.
Regarding other Close method calls - a defer command ignores any errors from a deferred function, and go-linter properly skips these... it's a normal practice and brings no value, so I'll leave the other places as is.

Reply:
Looking more I've noticed that there is an updateSecret function which after my changes does almost the same thing.
So, I'll combine them (literally, leave only one, whichever is simpler).